Add check on return value from start_client_batch

ncteisen opened the pull request:
The return value of client_start_batch was previously unchecked, which could cause a client to hang if it attempted to send a malformed header (see #8814).

There is now a check on the return of client_start_batch, and if there is a failure, then either an exception is raised, or an error rendezvous is returned depending on if the RPC was synchronous or asynchronous

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_channel.py:
```diff
@@ -99,6 +99,24 @@ def _wait_once_until(condition, until):
       condition.wait(timeout=remaining)
 
+def _check_call_error(call_error, metadata=None):
+  if call_error != cygrpc.CallError.ok:
```
Reduce the nesting by revising the control flow to be
```
if <one condition>:
  raise <one expression>
elif <another condition>:
  raise <another expression>
```
.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_channel.py:
```diff
@@ -99,6 +99,24 @@ def _wait_once_until(condition, until):
     else:
       condition.wait(timeout=remaining)
 
+def _check_call_error(call_error, metadata=None):
```
Keyword arguments generally are not appropriate for private methods - make `metadata` positional?

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_channel.py:
```diff
@@ -496,7 +515,11 @@ def future(self, request, timeout=None, metadata=None, credentials=None):
+            event_handler)
+        if call_error != cygrpc.CallError.ok:
+          return _Rendezvous(
+              _call_error_RPCstate(call_error, metadata), None, None, deadline)
```
Normally I'm the guy who's totally against all mutation, but... I think this code is easier to reason about if there remains exactly one `_RPCState` per RPC. I think you should change `_call_error_RPCstate` to mutate the `_RPCState` passed to it rather than create a new instance.

ncteisen commented:
Addressed comments on _channel.py and included my testing file in the commit.

The testing file is pretty hacked up at this point. I can make it cleaner, and move it to the unit test folder if it deserves to be there.

nathanielmanistaatgoogle commented:
This change is significant enough that it must have its own test coverage. Please clean the test and put it where it's supposed to be.

ncteisen commented:
Added a unit test for this new behavior

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_channel.py:
```diff
@@ -99,6 +99,22 @@ def _wait_once_until(condition, until):
       condition.wait(timeout=remaining)
 
+def _check_call_error(call_error, metadata):
+  if call_error == cygrpc.CallError.ok:
```
Why this rather than
```
if <one error condition>:
  raise <one error>
elif <another error condition>:
  raise <another error>
```
?

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_channel.py:
```diff
@@ -99,6 +99,22 @@ def _wait_once_until(condition, until):
+  elif call_error == cygrpc.CallError.invalid_metadata:
+    raise ValueError("metadata was invalid: %s" % metadata)
+  else:
+    raise ValueError('Internal gRPC CallError %d. Please report to '
```
Avoid implicit string concatenation whenever you can.

If our issue tracker URL is going to appear twice in program source, define a constant for it?

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_channel.py:
```diff
@@ -99,6 +99,22 @@ def _wait_once_until(condition, until):
+  if call_error == cygrpc.CallError.ok:
+    return
+  elif call_error == cygrpc.CallError.invalid_metadata:
+    raise ValueError("metadata was invalid: %s" % metadata)
```
Maintain consistent use of single-quote as the string delimiter throughout this file.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    metadata = (('InVaLiD', 'UnaryRequestBlockingUnaryResponse'),)
+    expected_error = ValueError("metadata was invalid: %s" % metadata)
+    multi_callable = _unary_unary_multi_callable(self._channel)
+    with self.assertRaises(ValueError) as cm:
```
Avoid abbreviations for the same reason that single-letter identifiers are forbidden by the style guide. This should probably be named `exception_context`.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+  def testUnaryRequestBlockingUnaryResponse(self):
+    request = b'\x07\x08'
+    metadata = (('InVaLiD', 'UnaryRequestBlockingUnaryResponse'),)
+    expected_error = ValueError("metadata was invalid: %s" % metadata)
```
You don't really use this ValueError instance except for the value of its `message` attribute. Why create it?

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    expected_error_details = "metadata was invalid: %s" % metadata
+    multi_callable = _unary_unary_multi_callable(self._channel)
+    with self.assertRaises(Exception) as cm:
+      response_future = multi_callable.future(request, metadata=metadata)
```
Pull this out of the `with` so that the test has the effect of asserting that execution of this line **does not** raise an exception.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    metadata = (('InVaLiD', 'UnaryRequestFutureUnaryResponse'),)
+    expected_error_details = "metadata was invalid: %s" % metadata
+    multi_callable = _unary_unary_multi_callable(self._channel)
+    with self.assertRaises(Exception) as cm:
```
No more specific exception class? `Exception` is very broad.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    with self.assertRaises(Exception) as cm:
+      response_future = multi_callable.future(request, metadata=metadata)
+      response = response_future.result()
+    self.assertEqual(cm.exception.details(), expected_error_details)
```
Please also assert that `response_future.code()`'s return value is as expected (and maybe `response_future.details()`?

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    multi_callable = _unary_unary_multi_callable(self._channel)
+    with self.assertRaises(ValueError) as cm:
+      response = multi_callable(request, metadata=metadata)
+    self.assertEqual(cm.exception.message, expected_error.message)
```
Please maintain the gRPC-Python-codebase-wide convention of `self.assertEqual(<expected>, <actual>)` over `self.assertEqual(<actual>, <expected>)`.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    request = b'\x07\x08'
+    metadata = (('InVaLiD', 'UnaryRequestBlockingUnaryResponse'),)
+    expected_error = ValueError("metadata was invalid: %s" % metadata)
+    multi_callable = _unary_unary_multi_callable(self._channel)
```
Kind of an arbitrary matter of taste but consider moving the instantiation of the MultiCallable objects to `setUp` rather than having one-per-test-method:
```
self._unary_unary = <...>
self._unary_stream = <...>
<...>
```
.

nathanielmanistaatgoogle commented:
Does the new test need to be recorded in `tests.json`?

ncteisen commented:
Addressed all comments. Added the test to tests.json

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_channel.py:
```diff
@@ -99,6 +99,22 @@ def _wait_once_until(condition, until):
     else:
       condition.wait(timeout=remaining)
 
+_INTERNAL_CALLERROR_MESSAGE = (
```
Underscore between `CALL` and `ERROR` (corresponding to and due to the way `cygrpc.CallError` is capitalized).

nathanielmanistaatgoogle replied:
Since this contains a `%` to be replaced it is a message format rather than a message. `_INTERNAL_CALL_ERROR_MESSAGE_FORMAT`?

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_channel.py:
```diff
@@ -99,6 +99,22 @@ def _wait_once_until(condition, until):
       condition.wait(timeout=remaining)
 
+_INTERNAL_CALLERROR_MESSAGE = (
+    'Internal gRPC CallError %d. ' +
```
And just plain old ordinary "call error" here, since what semantically matters in this user-visible (or developer-visible) string is not the specific `CallError` class name.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    request = b'\x07\x08'
+    metadata = (('InVaLiD', 'UnaryRequestBlockingUnaryResponse'),)
+    expected_error_details = "metadata was invalid: %s" % metadata
+    with self.assertRaises(ValueError) as exception_ctx:
```
Avoid abbreviations as well as initialisms and acronyms - this should be `exception_context`.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    metadata = (('InVaLiD', 'UnaryRequestBlockingUnaryResponse'),)
+    expected_error_details = "metadata was invalid: %s" % metadata
+    with self.assertRaises(ValueError) as exception_ctx:
+      response = self._unary_unary(request, metadata=metadata)
```
Did you try running `pylint`? Did it alert you to this local field being assigned but never accessed?

I'd say drop the field, but if you think the test is more clear with it included call it `unused_response`.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,169 @@
+    self.assertEqual(expected_error_details, exception_ctx.exception.message)
+
+  def testStreamRequestFutureUnaryResponse(self):
+    requests = tuple(b'\x07\x08' for _ in range(test_constants.STREAM_LENGTH))
```
Why not just
```py
request_iterator = (b'\x07\x08' for _ in range(test_constants.STREAM_LENGTH))
```
rather than these two lines? You should only need to create the sequence value if you need to iterate over its elements more than once.

ncteisen commented:
Addressed all comments. Now pylint only complains about the indentation and missing docstring, but the all the other unit tests don't have docstrings, so I assumed that was ok?

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,179 @@
+    expected_error_details = "metadata was invalid: %s" % metadata
+    response_iterator = self._unary_stream(request, metadata=metadata)
+    with self.assertRaises(grpc.RpcError) as exception_context:
+      tuple(response_iterator)
```
By changing this from `tuple` to `next` you again narrow the execution from which you are asserting that an exception must rise (from "an exception arises from consuming the entire stream of responses (including the end-of-stream indicator)" to "an exception arises from consuming the very first response in the response stream").

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_invalid_metadata_test.py:
```diff
@@ -0,0 +1,179 @@
+    expected_error_details = "metadata was invalid: %s" % metadata
+    response_iterator = self._stream_stream(request_iterator, metadata=metadata)
+    with self.assertRaises(grpc.RpcError) as exception_context:
+      tuple(response_iterator)
```
Here too.

ncteisen commented:
Done
